elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,105 @@
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-4360[WFCORE-4360]
+* https://issues.jboss.org/browse/WFLY-12215[WFLY-12215]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-677[EAP7-677]
+* https://issues.jboss.org/browse/WFLY-11101[WFLY-11101]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:jkasik@redhat.com[Jan Kasik]
+
+=== Testing By
+
+[ ] Engineering
+[x] QE
+
+=== Affected Projects or Components
+
+Utility classes may be required within the WildFly Elytron project and will be developed as the need is identified, however the primary integration will be performed within the WildFly Core project as this enhancment is predominantly in relation to management model expression resolution.
+
+Additionally all documentation lives within the WildFly project so updates will be applied there accordingly.
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+The general intent for this enhancement is to develop some form of configurable expression resolver.
+
+The expression resolver will reference a specific alias within a specific credential store, this credential will be used to decrypt the expression which was previously encrypted using this credential.  The expression resolver will also contain configuration such as the algorithm to use. 
+
+The expression resolver should support both generated keys and password based encryption.
+
+Using the credential store it will be possible to either add or generate suitable keys to be used for the encryption / decryption of values.  This will need to include: -
+
+ * The standalone elytron tool.
+ * The management operations exposed on the credential store.
+
+Where generated keys are used the tools and management operations should support exporting and importing the keys to allow them to be 'moved' between credential stores.  Where possible this should be achieved using specification defined formats.
+
+Functionality should be added to allow an administrator to encrypt a clear text String so that it can be stored as an expression, this will be added to: -
+
+ * The standalone elytron tool.
+ * New management operations, this may however be against the expression resolver not the credential store. 
+
+The domain mode relationships will need to considered further, where we used the PicketBox vault we defined the vault within the host.xml and used it's values in the domain.xml.  As the credential store can both be defined in the host.xml and the domain.xml it would make sense to make use of the one stored within that descriptor, however this could be impacted based on how the configurable expression resolver is defined.
```

Comment:
Do I understand it correctly that the problem is whether allow expressions in both `domain.xml` and `host xml` or is the issue something else?

To be honest, I am not certain that I fully understand how definition of expression resolver will affect the location where credential store is defined. Can you please explain it to me?

Reply:
When using the PicketBox Vault the Vault definition always lived in the host.xml and was used to handle expressions in both the host.xml and the domain.xml of that node.

The credential store on the other hand is now defined within the Elytron subsystem and is always used to resolve credentials in the co-located configuration i.e. the one in the host.xml will be used to obtain credentials for host.xml and the one in domain.xml will be used for credentials in domain.xml.

I suspect the latter would be preferable for consistency with the new configuration making use of subsystems - however the interaction between the host.xml and domain.xml handling is quite complex so when we did into this further it may throw up some additional complexities that may force our direction one way or the other.

Reply:
@darranl Do you already know which approach will be used? IMHO this analysis should contain final decision.

Reply:
Hi @darranl! Is there any update for this situation?

Reply:
@honza-kasik No there is no further update at the moment, I do not expect much activity before March.

Reply:
@darranl The host controller process is not allowed to resolve expressions based on any configuration in the profiles in domain.xml. Those profiles are only usable at runtime by servers.

I don't believe we have any cases of host controller subsystems (i.e. configured in host.xml) being used as part of the server configuration, and if we don't we should not make this the first such case.